### PR TITLE
Slightly simplify code of all backends

### PR DESCRIPTION
**Pull request by okurz**

### CODE REVIEW

File: backend/driver.pm [hunk 1]
```diff
@@ -156,14 +154,13 @@ sub _send_json {
...
-        return;
-    }
+    return undef if defined $rsp;
+    # this might have been closed by signal handler
+    no autodie 'close';
+    close($self->{backend_process}->channel_out);
+    $self->{backend_process}->channel_out(undef);
+    $self->{backend_process}->stop;
+    return;
     return $rsp->{rsp};
```
Comment (perlpunk) [inline backend/driver.pm:164]: This line is now unreachable

Comment (okurz) [reply backend/driver.pm]: that should explain the test failures :) Updated, thx.

File: backend/driver.pm [hunk 2]
```diff
@@ -156,15 +154,12 @@ sub _send_json {
...
-        $self->{backend_process}->stop;
-        return;
-    }
-    return $rsp->{rsp};
+    return $rsp->{rsp} if defined $rsp;
+    # this might have been closed by signal handler
+    no autodie 'close';
+    close($self->{backend_process}->channel_out);
+    $self->{backend_process}->channel_out(undef);
+    $self->{backend_process}->stop;
```
Comment (perlpunk) [inline backend/driver.pm:162]: I think this needs an explicit `return` now, as currently the return value of the `stop` method would be returned.

File: backend/generalhw.pm [hunk 3]
```diff
@@ -161,14 +156,13 @@ sub start_serial_grab {
...
-        die "exec failed $!";
-    }
+    return unless $self->{serialpid} == 0;
+    setpgrp 0, 0;
+    open(my $serial, '>',  $self->{serialfile});
+    open(STDOUT,     ">&", $serial);
+    open(STDERR,     ">&", $serial);
+    exec($self->get_cmd('GENERAL_HW_SOL_CMD'));
+    die "exec failed $!";
     return;
```
Comment (Martchus) [inline backend/generalhw.pm:166]: This `return` is now useless.

File: backend/generalhw.pm [hunk 4]
```diff
@@ -161,14 +156,13 @@ sub start_serial_grab {
...
     $self->{serialpid} = fork();
-    if ($self->{serialpid} == 0) {
-        setpgrp 0, 0;
-        open(my $serial, '>',  $self->{serialfile});
-        open(STDOUT,     ">&", $serial);
-        open(STDERR,     ">&", $serial);
-        exec($self->get_cmd('GENERAL_HW_SOL_CMD'));
-        die "exec failed $!";
-    }
+    return unless $self->{serialpid} == 0;
```
Comment (coolo) [inline backend/generalhw.pm:159]: unless == looks so python :)

Comment (okurz) [reply backend/generalhw.pm:159]: uh, and what would look "more perl-ish"?

File: backend/generalhw.pm [hunk 5]
```diff
@@ -24,7 +24,6 @@ use autodie ':all';
 
 use base 'backend::baseclass';
 
-use bmwqemu 'diag';
```
Comment (Amrysliu) [inline backend/generalhw.pm:27]: removing this line affects the using `bmwqemu::diag()` in line 84?

Comment (okurz) [reply backend/generalhw.pm]: good catch, thx.

Comment (okurz) [reply backend/generalhw.pm]: fixed

**Comment by Amrysliu** [comment]
LGTM

**Pull request merged**